v0/destinations/salesforce/transform.js

```diff
@@ -2,171 +2,169 @@ const get = require("get-value");
-  let targetEndpoint = `${authorizationData[1]}/services/data/v${SF_API_VERSION}/sobjects/Lead`;
+// Check for externalId field under context and look for probable Salesforce objects
+// We'll make separate requests for every Salesforce Object types present under externalIds
+//
+// Expected externalId map for Contact object:
+//
+// ------------------------
+// {
+//   "type": "Salesforce-Contact",
+//   "id": "0035g000001FaHfAAK"
+// }
+// ------------------------
+//
+// We'll use the Salesforce Object names by removing "Salesforce-" string from the type field
+//
+// Default Object type will be "Lead" for backward compatibility
+async function getSalesforceIdFromPayload(message, authorizationData) {
+  // define default map
+  const salesforceMaps = [];
+
+  // get externalId
+  const externalIds = get(message, "context.externalId");
+
+  // if externalIds are present look for type `Salesforce-`
+  if (externalIds) {
+    externalIds.forEach(extIdMap => {
+      const { type, id } = extIdMap;
+      if (type.includes("Salesforce")) {
+        salesforceMaps.push({
+          salesforceType: type.replace("Salesforce-", ""),
+          salesforceId: id
+        });
+      }
+    });
+  }
 
-  // check if the lead exists
-  // need to perform a parameterized search for this using email
-  const { email } = getFieldValueFromMessage(message, "traits");
+  // if nothing is present consider it as a Lead Object
+  // BACKWORD COMPATIBILITY
+  if (salesforceMaps.length === 0) {
+    // its a lead object. try to get lead object id using search query
+    // check if the lead exists
+    // need to perform a parameterized search for this using email
+    const { email } = getFieldValueFromMessage(message, "traits");
 
-  const leadQueryUrl = `${authorizationData[1]}/services/data/v${SF_API_VERSION}/parameterizedSearch/?q=${email}&sobject=Lead&Lead.fields=id`;
+    const leadQueryUrl = `${authorizationData.instanceUrl}/services/data/v${SF_API_VERSION}/parameterizedSearch/?q=${email}&sobject=Lead&Lead.fields=id`;
 
-  // request configuration will be conditional
-  // POST for create, PATCH for update
-  const leadQueryResponse = await axios.get(leadQueryUrl, {
-    headers: {
-      Authorization: authorizationData[0]
-    }
-  });
+    // request configuration will be conditional
+    const leadQueryResponse = await axios.get(leadQueryUrl, {
+      headers: { Authorization: authorizationData.token }
+    });
 
-  if (leadQueryResponse && leadQueryResponse.data.searchRecords) {
-    const retrievedLeadCount = leadQueryResponse.data.searchRecords.length;
-    // if count is greater than zero, it means that lead exists, then only update it
-    // else the original endpoint, which is the one for creation - can be used
-    if (retrievedLeadCount > 0) {
-      targetEndpoint += `/${leadQueryResponse.data.searchRecords[0].Id}?_HttpMethod=PATCH`;
+    let leadObjectId;
+    if (leadQueryResponse && leadQueryResponse.data.searchRecords) {
```

Comment:
check if leadQueryResponse.data is present

Reply:
Added the checking